PEPPER-1171_KitTrackingPKClashes

zyme opened the pull request:
DDP-1171 has a good amount of background, so take a look there before diving into this one.

The main change here was updating `KitDaoImpl` so that instead of preventing a duplicate row by waiting for the unique key to break on insert, it looks for a row that would break the unique key prior to insert.  In its previous form, because a unique key violation triggers a `SQLException` which in turn becomes an alert, which is a distraction for us because this a user error that does not require any intervention on our part.  What's important is that the user gets feedback about a duplicate scan, which still happens after this PR, but without any alerting and without any unique key issues.

Note that [PR to remove the `KitDao` interface](https://github.com/broadinstitute/ddp-study-server/pull/2739) will be merged into this branch *after* this PR is approved.  Since much is changing in `KitDaoImpl`, it's easier to keep these changes separate.

Likewise, a separate PR for renaming `ScanError` -> `ScanResult` is [here](https://github.com/broadinstitute/ddp-study-server/pull/2743) and will be merged after this PR is approved.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have added regression test labels to the jira ticket
- [ ] I have added verification steps to the jira ticket
- [ x] I have written automated positive tests
- [x ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

zyme committed:
PEPPER-1171 eliminate primary key collisions in KitTrackingScanUseCase

zyme committed:
PEPPER-1171 eliminate primary key collisions in KitTrackingScanUseCase

zyme committed:
PEPPER-1171 eliminate primary key collisions in KitTrackingScanUseCase

zyme committed:
Merge branch 'develop' into PEPPER-1171_KitTrackingPKClashes

zyme committed:
PEPPER-1171 restoring previous insert...set syntax

zyme committed:
PEPPER-1171 WIP

zyme committed:
Merge branch 'develop' into PEPPER-1171_KitTrackingPKClashes

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java

zyme committed:
PEPPER-1171 WIP

zyme committed:
Merge branch 'develop' into PEPPER-1171_KitTrackingPKClashes

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -530,30 +574,58 @@ public KitRequestShipping getKitRequestShipping(@NonNull ResultSet rs) throws SQ
-                    dbVals.resultValue = new ScanError(kitRequestShipping.getKitLabel(),
-                            String.format("Kit Label \"%s\" does not exist.\n"
-                                    + UserErrorMessages.IF_QUESTIONS_CONTACT_DEVELOPER, kitRequestShipping.getKitLabel()));
-                    logger.error(String.format("The number of affected rows for kit label %s is not 1", kitRequestShipping.getKitLabel()));
```
The change here is that instead of erroring out when > 1 row is updated, we check for the row first and return a scan error without breaking the constraint and sending an alert.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/KitTrackingScanUseCase.java:
```diff
@@ -17,22 +19,8 @@ public KitTrackingScanUseCase(KitPayload kitPayload, KitDao kitDao) {
 
     @Override
     protected Optional<ScanError> process(ScanPayload scanPayload) {
-        String kitLabel = scanPayload.getKitLabel();
```
This logic has moved into the dao, so that we can test more code without having to hit the route.

zyme committed:
PEPPER-1171 reverting long -> int since the id is indeed an int.

zyme committed:
PEPPER-1171 WIP

zyme committed:
PEPPER-1171 WIP

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -191,7 +190,7 @@ public class KitRequestShipping extends KitRequest implements HasDdpInstanceId {
             " SELECT count(*) kitRequestCount from ddp_kit_request where bsp_collaborator_sample_id REGEXP \"^%1\" and kit_type_id = ?";
 
-    private static final String GET_FOUND_IF_KIT_WITH_DDP_LABEL_ALREADY_EXISTS =
+    public static final String SELECT_1_FROM_KIT_REQUEST_WITH_DDP_LABEL =
```
Needed to run this query from another dao, so I made it public.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1243,7 +1243,7 @@ public static int getKitCounter(@NonNull Connection conn, String collaboratorSam
      */
     public static String generateDdpLabelID() {
-        return generateDdpLabelID(15, GET_FOUND_IF_KIT_WITH_DDP_LABEL_ALREADY_EXISTS);
+        return generateDdpLabelID(15, SELECT_1_FROM_KIT_REQUEST_WITH_DDP_LABEL);
```
Renamed this to make it a bit more clear what the query does.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1748,4 +1748,9 @@ public boolean isKitRequiringTrackingScan() {
+
+    // todo arz test me
+    public boolean hasBSPCollaboratorParticipantId() {
+        return StringUtils.isNotBlank(bspCollaboratorParticipantId);
```
I saw the check for blank bsp collaborator participant id in a few places so I extracted a method for it.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -530,30 +646,59 @@ public KitRequestShipping getKitRequestShipping(@NonNull ResultSet rs) throws SQ
                 if (rowsAffected != 1) {
-                    dbVals.resultValue = new ScanError(kitRequestShipping.getKitLabel(),
-                            String.format("Kit Label \"%s\" does not exist.\n"
-                                    + UserErrorMessages.IF_QUESTIONS_CONTACT_DEVELOPER, kitRequestShipping.getKitLabel()));
```
Instead of having "is not 1", expanded the queries to help me understand whether 1 row or 0 rows is expected in different situations.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -530,30 +646,59 @@ public KitRequestShipping getKitRequestShipping(@NonNull ResultSet rs) throws SQ
+                    return dbVals;
                 }
-            } catch (Exception ex) {
-                dbVals.resultValue = new ScanError(kitRequestShipping.getKitLabel(),
```
Rather than catching `Exception` and issuing an alert, this now catches only `SQLException` so that we can return a ScanError for some situations without sending an alert.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/BaseKitUseCase.java:
```diff
@@ -37,23 +33,18 @@ public List<ScanError> get() {
-    }
-
-
-    protected BaseKitUseCase getDecoratedScanUseCase() {
```
Unused method, so I removed it.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/BaseKitUseCase.java:
```diff
@@ -37,23 +33,18 @@ public List<ScanError> get() {
-        }
-        return decoratedScanUseCase;
+        return maybeScanError.isEmpty()
+                || maybeScanError.get().isScanErrorOnlyBspParticipantId(bspCollaboratorParticipantId);
```
Little bit of checkstyle clean up.

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/kit/ScanError.java:
```diff
@@ -1,27 +1,39 @@
 
 import lombok.Data;
-import spark.utils.StringUtils;
+import org.apache.commons.lang3.StringUtils;
```
Wrong import--this should be directly from apache.

zyme committed:
PEPPER-1171 WIP

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -163,32 +196,98 @@ public Boolean hasTrackingScan(String kitLabel) {
+                            "kit " + kitRequestShipping.getDdpLabel() + " not found.");
+                    return dbVals;
+                } else if (numKitsWithLabel > 1) {
+                    // todo arz is this an error?  Expected to have multiple kits with the same label?
```
@pegahtah can you confirm that this should indeed be an error?

pegahtah replied:
Yes, I double checked, ddp_label should be unique per kit entry

zyme commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -163,32 +196,98 @@ public Boolean hasTrackingScan(String kitLabel) {
+                    logger.info("Updated ddp_kit.kit_label to {} for kit request  {}.",
+                            kitRequestShipping.getKitLabel(), kitRequestShipping.getDdpLabel());
+                    if (kitRequestShipping.hasBSPCollaboratorParticipantId()) {
+                        // todo arz how does this work?  A scan error with no error?
```
@pegahtah can you help me understand this special block of code that shows up when there's a bsp collab participant id?

pegahtah replied:
Of course, for understanding this I need to first tell you  that the ScanError is used to communicate anything (not just errors) with the frontend. If the kit has a bsp participant id, we will send it to the frontend. So the frontend will show the successful scan using the returned response. (GP wants to see that as a final step to make sure they are sending the kit to the right participant)
 As an example Angio Dev returned this `[{"kit":"W9O6M23K73T36N8","shortId":"ASCProject_P9846A"}]` for a successful scan, in which `kit` is the `DSM shipping label` or `ddpLabel` and frontend uses it to display the bspParticipantId under the correct corresponding scan line
![Screenshot 2023-11-22 at 11 09 45 AM](https://github.com/broadinstitute/ddp-study-server/assets/47120870/1abdba87-dff7-45f4-a6b9-11fb65c65a2f)

zyme replied:
What would you say to changing `ScanError` -> `ScanResponse` to reflect the fact that it's not just for errors?

zyme replied:
PR for *just* changing ScanError -> ScanResult is [here](https://github.com/broadinstitute/ddp-study-server/pull/2743).  Keeping it separate so we can evaluate whether we want to do it independently.

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -612,6 +757,39 @@ public List<KitRequestShipping> getKitsByHruid(String hruid) {
         return kitRequestList;
     }
 
+    public void updateKitRequestDDPLabel(int dsmKitRequestId, String ddpLabel) {
```
I am not able to find the usage of this method, but in general I can't think of the use case for changing the ddpLabel of a kit. It is one of the unique ids of the kit

zyme replied:
Gaah, I forgot to remove this--was a leftover from an earlier approach.  Will delete.

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -612,6 +757,39 @@ public List<KitRequestShipping> getKitsByHruid(String hruid) {
+        });
+    }
+
+    public void updateKitComplete(long dsmKitId, boolean isKitComplete) {
```
Can this be private and `VisibleForTesting` maybe?

zyme committed:
DDP-1171 PR feedback

zyme committed:
Merge branch 'develop' into PEPPER-1171_KitTrackingPKClashes

zyme committed:
PEPPER-1205 added script to hard delete previously soft-deleted onc histories, etc. (#2741)

PEPPER-1205 added script to hard delete previously soft-deleted sm ids, tissues, and onc histories.

zyme committed:
PEPPER-655 Use VPC egress instead of docker (#2746)

PEPPER-655 added VPC egress connector to appengine config

zyme committed:
PEPPER-655 WIP

pegahtah committed:
[DSM] PEPPER-1059 (#2737)

* PEPPER-1059 made first test that is not passing

* PEPPER-1059 made first test that is not passing

* PEPPER-1059 Changed BaseMigrator.java to get the guids of all participants in the study before export to make sure deleted data gets acknowledged and is overwritten in export

* PEPPER-1059 fixed the logging

* PEPPER-1059 fixed the try-catch

* PEPPER-1059 changed the `deleteOncHistoryDirectlyFromDB` to call Dao

* PEPPER-1059 checkstyle fix

* PEPPER-1059 checkstyle fixes (may introduce some noise into PR )

* PEPPER-1059 one more test, PR comments applied and added retries for clearing scroll

* PEPPER-1059 PR comments and a javadoc comment added

* PEPPER-1059 Changing the getAllPartici

* PEPPER-1059 changed the scope of the `participantsInTheStudy` to be in the method and added one more test

* PEPPER-1059 PR feedback

zyme committed:
PEPPER-655 WIP

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -66,19 +68,31 @@ public class KitDaoImpl implements KitDao {
             + "WHERE tracking.kit_label = ?) AS existing_rows";
-    private static final String UPDATE_KIT_REQUEST = "UPDATE ddp_kit SET "
+
+    private static final String UPDATE_KIT_REQUEST_DDP_LABEL = "update ddp_kit_request set ddp_label = ? where "
```
Now that the method is removed do we need this one?

zyme replied:
Indeed we don't

zyme committed:
PEPPER-655 WIP fake commit to trigger circle

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -163,32 +196,98 @@ public Boolean hasTrackingScan(String kitLabel) {
+                        "An unexpected error has occurred");
+            }
+
+            try (PreparedStatement stmt = conn.prepareStatement(SQL_NUM_KITS_BY_LABEL)) {
```
Do we need this after checking the ddpLabel exists in Line 211 with `SELECT_1_FROM_KIT_REQUEST_WITH_DDP_LABEL`?  Maybe we can do only `SQL_NUM_KITS_BY_LABEL` which is checking both ddp_kit and ddp_kit_request table?

zyme replied:
I think we need both?  One checks for existence of the kit request, the other for the ddp kit.

zyme replied:
On the other hand, all the tests pass after I remove this query, so I think we can ditch it!

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -163,32 +196,98 @@ public Boolean hasTrackingScan(String kitLabel) {
+            } catch (SQLException ex) {
+                logger.error("Not able to query the kit request for ddpLabel " + kitRequestShipping.getDdpLabel(), ex);
+                dbVals.resultValue = new ScanError(kitRequestShipping.getDdpLabel(),
+                        "An unexpected error has occurred");
```
Should we return the dbVals here? Or do we want to continue after a SQL exception?

zyme replied:
Nice catch!  Yes, this should return early.

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -369,14 +468,31 @@ public Integer deleteKitRequest(Long kitRequestId) {
         if (simpleResult.resultException != null) {
-            throw new DsmInternalError(String.format("Error deleting kit request with id: %d", kitRequestId),
-                    simpleResult.resultException);
+            throw new RuntimeException("Error deleting kit request with id: " + kitRequestId, simpleResult.resultException);
```
I think DSMInternalError is more accurate for here since we are moving on from RuntimeExcpetions

zyme committed:
PEPPER-655 PR feedback

zyme committed:
PEPPER-1171 PR feedback

zyme committed:
Pepper 1171 remove kit dao interface (#2749)

PEPPER-1171 Removed KitDao interface, replacing it with KitDaoImpl.

zyme committed:
PEPPER-1171 Rename ScanError to ScanResult (#2743)

DDP-1171 changing ScanError to ScanResult

zyme committed:
Merge branch 'develop' into PEPPER-1171_KitTrackingPKClashes

zyme committed:
PEPPER-1171 WIP

zyme committed:
Merge branch 'develop' into PEPPER-1171_KitTrackingPKClashes

zyme committed:
Merge branch 'develop' into PEPPER-1171_KitTrackingPKClashes
